// uppy-store-ngrx/dist/src/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Subject_1 = require("rxjs/Subject");
var utils_1 = require("./utils");
var StoreActions = require("./actions");
var DEFAULT_STATE_ID = 'uppy';
// Pluck Uppy state from the Ngrx store in the default location.
var defaultSelector = function (id) { return function (state) { return state[id]; }; };
/**
 * T - interface of root state;
 * U - interface of custom file;
 */
var NgrxStore = /** @class */ (function () {
    function NgrxStore(opts) {
        this._store = opts.store;
        this._id = opts.id || DEFAULT_STATE_ID;
        this._selector = opts.selector || defaultSelector(this._id);
        // Initialise the `uppy[id]` state key.
        this.setState({});
    }
    NgrxStore.prototype.setState = function (patch) {
        this._store.dispatch(new StoreActions.StateUpdate({
            patch: patch,
            id: this._id,
        }));
    };
    NgrxStore.prototype.getState = function () {
        var state;
        this._store.select(this._selector).take(1).subscribe(function (s) { return state = s; });
        return state;
    };
    NgrxStore.prototype.subscribe = function (cb) {
        var unsub$ = new Subject_1.Subject();
        var prevState = this.getState();
        this._store.takeUntil(unsub$).subscribe(function (nextState) {
            if (prevState !== nextState) {
                var patch = utils_1.getPatch(prevState, nextState);
                cb(prevState, nextState, patch);
                prevState = nextState;
            }
        });
        return function () { return unsub$.next(true); };
    };
    NgrxStore.prototype.getStore = function () {
        return this._store.select(this._selector);
    };
    NgrxStore.prototype.getFiles = function () {
        return this.getStore().select(function (state) { return state.files; });
    };
    return NgrxStore;
}());
exports.NgrxStore = NgrxStore;
function createNgrxStore(opts) {
    return new NgrxStore(opts);
}
exports.createNgrxStore = createNgrxStore;
